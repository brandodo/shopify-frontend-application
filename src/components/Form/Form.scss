.form {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: 40%;
  padding: 20px;

  &__container {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 15px;
    padding-bottom: 15px;
  }

  &__label {
    font-weight: 600;
    font-size: 36px;
    letter-spacing: 2px;
    color: rgb(135, 0, 0);
  }

  &__input {
    width: 350px;
    text-align: center;
    border-radius: 50px;
    font-size: 44px;
  }

  &__options {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    gap: 15px;
    width: 100%;
    padding-bottom: 15px;
  }

  &__select {
    width: 250px;
    border-radius: 50px;
    border: 2px solid black;
    font-size: 44px;
    text-align: center;
  }

  &__avatar {
    width: 100px;
  }

  &__submit {
    width: 250px;
    padding: 10px;
    border-radius: 50px;
    background-color: rgb(135, 0, 0);
    color: #fff;
    font-size: 32px;
    transition: all 0.25s ease;

    &:hover {
      cursor: pointer;
      background-color: black;
      border: 2px solid#fff;
    }
  }
}
