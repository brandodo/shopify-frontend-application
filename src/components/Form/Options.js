import davinci from "../../assets/images/davinci.png";
import curie from "../../assets/images/curie.png";
import babbage from "../../assets/images/babbage.png";
import ada from "../../assets/images/ada.png";
import { useEffect } from "react";

export const Options = ({ engine, setEngine, avatar, setAvatar }) => {
  useEffect(() => {
    switch (engine) {
      case "text-davinci-002":
        setAvatar(davinci);
        break;

      case "text-curie-001":
        setAvatar(curie);
        break;

      case "text-babbage-001":
        setAvatar(babbage);
        break;

      case "text-ada-001":
        setAvatar(ada);
        break;

      default:
        break;
    }
  }, [engine]);

  return (
    <div className="form__options">
      <div className="form__container">
        <label className="form__label" htmlFor="engine">
          Choose your curator
        </label>
        <select
          className="form__select"
          name="engine"
          value={engine}
          onChange={(event) => {
            setEngine(event.target.value);
          }}
        >
          <option className="form__option" value="text-davinci-002">
            DaVinci
          </option>
          <option className="form__option" value="text-curie-001">
            Madam Curie
          </option>
          <option className="form__option" value="text-babbage-001">
            Count Babbage
          </option>
          <option className="form__option" value="text-ada-001">
            Apparition Ada
          </option>
        </select>
      </div>
      <img className="form__avatar" src={avatar} alt="curator-avatar" />
    </div>
  );
};
