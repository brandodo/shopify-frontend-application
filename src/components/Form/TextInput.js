export const TextInput = ({ setInput, label }) => {
  return (
    <div className="form__container">
      <label className="form__label" htmlFor="text-input">
        {label}
      </label>
      <input
        name="text-input"
        className="form__input"
        type="text"
        onChange={(event) => setInput(event.target.value)}
      />
    </div>
  );
};
