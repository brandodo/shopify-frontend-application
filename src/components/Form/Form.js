import { useState } from "react";
import "./Form.scss";
import { Options } from "./Options";
import { TextInput } from "./TextInput";

export const Form = ({ formHandle }) => {
  const [input, setInput] = useState("");
  const [engine, setEngine] = useState("text-curie-001");
  const [avatar, setAvatar] = useState("");

  const disableButton = input ? false : true;

  return (
    <form
      className="form"
      onSubmit={(event) => {
        event.preventDefault();
        formHandle(input, engine, avatar);
      }}
    >
      <TextInput setInput={setInput} label="Write a thriller about" />
      <Options
        engine={engine}
        setEngine={setEngine}
        setAvatar={setAvatar}
        avatar={avatar}
      />
      <input
        className={
          disableButton
            ? "form__submit form__submit--disabled"
            : "form__submit form__submit--hover"
        }
        name="submit-button"
        type="submit"
        disabled={disableButton}
      />
    </form>
  );
};
