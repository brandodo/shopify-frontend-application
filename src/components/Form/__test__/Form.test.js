import React from "react";
import { Form } from "../Form";
import {
  render,
  cleanup,
  waitFor,
  screen,
  within,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import user from "@testing-library/user-event";

afterEach(cleanup);

describe("<Form />", () => {
  // test # 1
  it("renders without error", () => {
    let component = renderer.create(<Form />);
    expect(component).toBeTruthy;
  });

  // test # 2
  it("matches snapshot", () => {
    let component = renderer.create(<Form />);
    expect(component).toMatchSnapshot();
  });

  const onSubmit = jest.fn();

  beforeEach(() => {
    onSubmit.mockClear();
    render(<Form formHandle={onSubmit} />);
  });

  // test # 3
  it("onSubmit to be called when input filled", async () => {
    user.type(getTextInput(), "Spaghetti");
    user.click(getButton());

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalledWith(
        "Spaghetti",
        "text-curie-001",
        "curie.png"
      );
    });

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });

  // test # 4
  it("onSubmit not called when input empty regardless of selected option", () => {
    user.click(getButton());

    const dropdown = getDropdown();
    user.selectOptions(
      dropdown,
      within(dropdown).getByRole("option", { name: /count babbage/i })
    );

    expect(onSubmit).not.toHaveBeenCalled();
  });

  // test # 5
  it("onSubmit called with input filled and different dropdown option", async () => {
    user.type(getTextInput(), "dinosaurs");

    const dropdown = getDropdown();
    user.selectOptions(
      dropdown,
      within(dropdown).getByRole("option", { name: /davinci/i })
    );

    user.click(getButton());

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalledWith(
        "dinosaurs",
        "text-davinci-002",
        "davinci.png"
      );
    });

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});

const getTextInput = () => {
  return screen.getByRole("textbox");
};

const getDropdown = () => {
  return screen.getByRole("combobox");
};

const getButton = () => {
  return screen.getByRole("button", {
    name: /submit/i,
  });
};
