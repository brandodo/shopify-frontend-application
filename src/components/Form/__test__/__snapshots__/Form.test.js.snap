// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> matches snapshot 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__container"
  >
    <label
      className="form__label"
      htmlFor="text-input"
    >
      Write a thriller about
    </label>
    <input
      className="form__input"
      name="text-input"
      onChange={[Function]}
      placeholder="e.g. Pizza"
      type="text"
    />
  </div>
  <div
    className="form__options"
  >
    <div
      className="form__container"
    >
      <label
        className="form__label"
        htmlFor="engine"
      >
        Choose your curator
      </label>
      <select
        className="form__select"
        name="engine"
        onChange={[Function]}
        value="text-curie-001"
      >
        <option
          className="form__option"
          value="text-davinci-002"
        >
          DaVinci
        </option>
        <option
          className="form__option"
          value="text-curie-001"
        >
          Madam Curie
        </option>
        <option
          className="form__option"
          value="text-babbage-001"
        >
          Count Babbage
        </option>
        <option
          className="form__option"
          value="text-ada-001"
        >
          Apparition Ada
        </option>
      </select>
    </div>
    <img
      alt="curator-avatar"
      className="form__avatar"
      src="curie.png"
    />
  </div>
  <input
    className="form__submit form__submit--disabled"
    disabled={true}
    name="submit-button"
    type="submit"
  />
</form>
`;
